jwst/extract_1d/ifu.py

```diff
@@ -384,14 +413,107 @@ def extract_ifu(input_model, source_type, extract_params):
+        X and Y coordinates (in that order) of the pixel that has right
+        ascension and declination coordinates closest to the target
+        location.  The spectral extraction region should be centered here.
+    """
+
+    if ra_targ is None or dec_targ is None:
+        log.warning("TARG_RA and/or TARG_DEC not found; can't compute "
+                    "pixel location of target.")
+        locn = None
+    else:
+        shape = input_model.data.shape
+        grid = np.indices(shape[-2:])
+        z = np.zeros(shape[-2:], dtype=np.float64) + shape[0] // 2
+        # The arguments are the X, Y, and Z pixel coordinates, and the
+        # output arrays will be 2-D.
+        (ra_i, dec_i, wl) = input_model.meta.wcs(grid[1], grid[0], z)
+        rect = sph_to_rect(ra_i, dec_i)
+        v = sph_to_rect(ra_targ, dec_targ)
+        diff = rect - v
+        # We want the pixel with the minimum distance from v, but the pixel
+        # with the minimum value of distance squared will be the same.
+        dist2 = (diff**2).sum(axis=-1)
+        nan_mask = np.isnan(wl)
+        dist2[..., :] = np.where(nan_mask, HUGE_DIST, dist2[..., :])
+        del nan_mask
+        k = np.argmin(dist2.ravel())
+        (j, i) = divmod(k, dist2.shape[1])      # y, x coordinates
+
+        if i <= 0 or j <= 0 or i >= shape[-1] - 1 or j >= shape[-2] - 1:
+            log.warning("WCS implies the target is at or beyond the edge "
+                        "of the image; this location will not be used.")
+            locn = None
+        else:
+            locn = (i, j)                       # x, y coordinates
+
+    return locn
+
+
+def sph_to_rect(ra, dec):
+    """Convert celestial coordinates to rectangular.
+
+    Parameters
+    ----------
+    ra, dec : ndarray or float
+        The right ascension and declination (degrees).  Both `ra` and `dec`
+        should be arrays of the same shape, or they should both be float.
+
+    Returns
+    -------
+    rect : ndarray
+        If `ra` and `dec` are float, `rect` with be a 3-element array.
+        If `ra` and `dec` are arrays, `rect` will be an array with shape
+        ra.shape + (3,).
+        For each element of `ra` (or `dec`), the last axis of `rect` will
+        give the rectangular coordinates of a unit vector in the direction
+        `ra`, `dec`.  The elements of the vector in rectangular coordinates
+        are in the order x, y, z, where x is the direction toward the
+        vernal equinox, y is the direction toward right ascension = 90
+        degrees (6 hours) and declination = 0, and z is toward the north
+        celestial pole.
+    """
+
+    if hasattr(ra, 'shape'):
+        shape = ra.shape + (3,)
+    else:
+        shape = (3,)
+
+    rect = np.zeros(shape, dtype=np.float64)
+    rect[..., 2] = np.sin(dec * np.pi / 180.)
+    r_xy = np.cos(dec * np.pi / 180.)
+    rect[..., 1] = r_xy * np.sin(ra * np.pi / 180.)
+    rect[..., 0] = r_xy * np.cos(ra * np.pi / 180.)
+
+    return rect
```

Comment:
We have this kind of computation in several places. Should probably go in a library.

Reply:
I agree.  I wrote this because I knew what I wanted (the input can be either floats or arrays), and it was easy.